contracts/interchain-token-service/src/contract/execute.rs

```diff
@@ -186,16 +247,168 @@ pub fn set_chain_config(
+    let mut destination_instance = try_load_token_instance(
+        storage,
+        destination_chain.clone(),
+        transfer.token_id.clone(),
+    )?;
+
+    TokenInstance::update_supply(
+        &mut source_instance,
+        &mut destination_instance,
+        transfer.amount.clone(),
+    )
+    .change_context_lazy(|| Error::TokenSupplyInvariantViolated {
+        token_id: transfer.token_id.clone(),
+        chain: source_chain.clone(),
+    })?;
+
+    state::save_token_instantiation(
+        storage,
+        source_chain,
+        transfer.token_id.clone(),
+        &source_instance,
+    )
+    .and_then(|_| {
+        state::save_token_instantiation(
+            storage,
+            destination_chain,
+            transfer.token_id.clone(),
+            &destination_instance,
+        )
+    })
+    .change_context(Error::State)
+}
+
+fn try_load_token_instance(
+    storage: &mut dyn Storage,
+    chain: ChainNameRaw,
+    token_id: TokenId,
+) -> Result<TokenInstance, Error> {
+    state::may_load_token_instance(storage, chain.clone(), token_id.clone())
+        .change_context(Error::State)?
+        .ok_or(report!(Error::TokenNotDeployed {
+            token_id: token_id.clone(),
+            chain,
+        }))
+}
+
+fn apply_token_deployment(
+    storage: &mut dyn Storage,
+    source_chain: ChainNameRaw,
+    destination_chain: ChainNameRaw,
+    token_id: TokenId,
+    deployment_type: &TokenDeploymentType,
+) -> Result<(), Error> {
+    ensure_token_not_deployed_on_destination(storage, token_id.clone(), destination_chain.clone())?;
+
+    let token_config =
+        state::may_load_global_token_config(storage, &token_id).change_context(Error::State)?;
+
+    if let Some(GlobalTokenConfig { origin_chain, .. }) = token_config {
+        ensure_origin_matches_source_chain(source_chain, origin_chain, token_id.clone())?;
+    } else {
+        initialize_token_on_origin(storage, source_chain, token_id.clone())?;
+    }
+
+    let instance = TokenInstance::new(token_id.clone(), deployment_type);
+
+    state::save_token_instantiation(storage, destination_chain, token_id, &instance)
+        .change_context(Error::State)
+}
+
+fn ensure_origin_matches_source_chain(
+    source_chain: ChainNameRaw,
+    origin_chain: ChainNameRaw,
+    token_id: TokenId,
+) -> Result<(), Error> {
+    ensure!(
+        origin_chain == source_chain,
+        Error::TokenDeployedFromNonOriginChain {
+            token_id,
+            origin_chain,
+            chain: source_chain.clone(),
+        }
+    );
+
+    Ok(())
+}
+
+fn initialize_token_on_origin(
+    storage: &mut dyn Storage,
+    source_chain: ChainNameRaw,
+    token_id: TokenId,
+) -> Result<(), Error> {
+    // Token is being deployed for the first time
+    let token_config = GlobalTokenConfig {
+        token_id: token_id.clone(),
+        origin_chain: source_chain.clone(),
+    };
+    let instance = TokenInstance::new_on_origin(token_id.clone());
```

Comment:
Why make the `GlobalTokenConfig` contain `token_id`? If there's a reason, why doesn't `TokenInstance` contain the `origin_chain`?

Reply:
Also, why is the `TokenInstance` on the original chain always untracked?

Reply:
I'll remove the token_ids again. I'm not really happy with what these structs contain, but for now I guess it's good enough to get it across the finish line with all the necessary data available and then we can adjust later.

Reply:
I guess the idea is that the original chain can always mint tokens and ITS doesn't have control over the token management